aws-s3-transfer-manager/src/operation/download/handle.rs

```diff
@@ -1,34 +1,56 @@
+use std::sync::Arc;
+
+use crate::error::{self, ErrorKind};
+use tokio::{
+    sync::{oneshot::Receiver, Mutex, OnceCell},
+    task,
+};
+
 /*
  * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 use crate::operation::download::body::Body;
-use crate::operation::download::object_meta::ObjectMetadata;
-use tokio::task;
 
-use super::DownloadContext;
+use super::object_meta::ObjectMetadata;
 
 /// Response type for a single download object request.
 #[derive(Debug)]
 #[non_exhaustive]
 pub struct DownloadHandle {
-    /// Object metadata
-    pub object_meta: ObjectMetadata,
+    /// Object metadata receiver.
+    pub(crate) object_meta_receiver: Mutex<Option<Receiver<ObjectMetadata>>>,
+    /// Object metadata.
+    pub(crate) object_meta: OnceCell<ObjectMetadata>,
 
     /// The object content
     pub(crate) body: Body,
 
-    /// All child tasks spawned for this download
-    pub(crate) tasks: task::JoinSet<()>,
+    /// Discovery task
+    pub(crate) discovery: task::JoinHandle<Result<(), error::Error>>,
 
-    /// The context used to drive an upload to completion
-    pub(crate) ctx: DownloadContext,
+    /// All child tasks (ranged GetObject) spawned for this download
+    pub(crate) tasks: Arc<Mutex<task::JoinSet<()>>>,
 }
 
 impl DownloadHandle {
     /// Object metadata
-    pub fn object_meta(&self) -> &ObjectMetadata {
-        &self.object_meta
+    pub async fn object_meta(&self) -> Result<&ObjectMetadata, error::Error> {
```

Comment:
my gut tells me we'd be better off with an API that mirrors what we're doing under the hood, and we can optionally wrap that with a "nice" API that hides certain details we don't think basic users need

So: Our API has an iterator, that iterates over each request/response in the operation, whether or not it has a body.
 * So if we did a HeadObject, the first item is HeadObject.
     * If Content-Length was 0, there will be no more items
     * Otherwise, there will be 1+ GetObject items afterwards.
 * If we do discovery via GetObject, then there will simply be 1+ GetObject items.
 * It is what it its

We can wrap that in a "nice" API where there are different getters for "Object Metadata" vs body chunks, and in that wrapper we can deal with the complexity of trying to differentiate the discovery request from the first-chunk request, and whether or not they're actually the same request.

Just imagine the headache our advanced users are going to have trying to "unwrap" the complexity we bake into the base API. If they're logging metrics from the stream of HTTP calls, they'd need different branches for discover vs body chunks, and they'd need to figure out whether the 1st chunk request was actually the same as the discovery request.

Reply:
> they'd need to figure out whether the 1st chunk request was actually the same as the discovery request.

With the current API, it's straightforward because the object metadata will contain `request_id` only if we perform a `HeadObject` and not a `GetObject`. The `GetObject` will always be delivered from the body iterator.

> Our API has an iterator, that iterates over each request/response in the operation, whether or not it has a body.

I think the current design more closely mirrors what we are actually doing, which is a separate/specific discovery request followed by N parallel `GetObject` operations. I thought about the iterator-based approach but didn't implement it due to the following considerations:
- I think it won't be what our users expect. They will expect N part-size chunks, and receiving a first chunk that is just a `HeadObject` is not that interesting.
- In the future, we might just get rid of `HeadObject` and perform a 1-byte `GetObject` request to retrieve the metadata (to support presign cases, etc.). In that case, do we simply expose that metadata as it is? That would be confusing for customers.
- What about the empty file use case? Will we deliver two requests with empty bodies?

An iterator seems nice and simple, but I think it will make hiding the complexity harder as we go along. Advanced users should have the ability to get the information they need, but we also need the ability to hide certain complexities from the users. 

I think the approach of if we make any request solely for discovery, we will deliver its metadata in `object_meta` and we will deliver N `GetObject` requests with their metadata, is a nice compromise. It lets us hide the complexity of discovery, and we can change it as needed in the future. I'd like to hear what others think about this.

Reply:
Yeah, might be a mistake confusing metadata and metrics/telemetry.

Metadata is for a customer that needs  to know some detail about a request/response that we didn't know they were interested in (e.g. part checksums).

Metrics/telemetry would include stuff like failed requests that got retried. And the use case for this is customers needing to look under the hood to diagnose bugs and performance issues (e.g. https://github.com/awslabs/mountpoint-s3/issues/1079).

Reply:
Maybe we don't even want to call this metadata, it's all the fields that can be set/returned on an object that isn't the object content itself. In SDKs this is all just part of e.g. `GetObjectResponse` but here we have to give it a name because we aren't exposing it the same way.

Reply:
Discussed offline. Currently, we will be moving forward with the current API. We will consider an iterator-like API for telemetry in the future, which will require us to hook into the SDK's interceptor so that we can get all the requests, including retries, etc., and expose that.